userchroot.c

```diff
@@ -284,7 +284,6 @@ static int mount_shm(void) {
     mode_t perms = (0777 | S_ISVTX);
     const char *path = "/dev/shm";
 
-    mkdir(path, perms);
     if (chown(path, 0, 0) < 0)
```

Comment:
We need to remove this `chown()`, otherwise the directory is always owned by root and the `st_uid` check is pointless. The top-level directory of the tmpfs will still be owned by root after mounting, independent of the owner of the underlying mount point.

I can't comment on the unmodified `chmod()` line further down, so commenting here: We should remove the call to `chmod()` as well. It doesn't affect the default mode of the top-level directory of the tmpfs and it's better not to give arbitrary users write access to the unmounted directory. The default mode of the top-level tmpfs directory is 1777, which is what we want.

Reply:
You're totally right. Looks like a lot of this logic is unnecessary, the permissions are correct once mounted. I've gone ahead and relaxed these restrictions, and also used `lstat` instead of `stat` to make sure it's a directory and not a symlink to something like a `root` owned directory.